crates/pairwise_nostd_internal/src/bins.rs

```diff
@@ -0,0 +1,194 @@
+//! bin index for a given value.
+
+/// Super simple. This can be expanded as needed.
+pub trait BinEdges {
+    /// Calculate the bin index for a given value. Values which are equal to
+    /// boundary values are considered part of the higher bin.
+    fn bin_index(&self, value: f64) -> Option<usize>;
+
+    fn len(&self) -> usize;
+
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+/// Regular bins with uniform spacing
+#[derive(Clone)]
+pub struct RegularBinEdges {
+    min: f64,
+    max: f64,
+    bin_size: f64,
+    n_bins: usize,
+}
+impl RegularBinEdges {
+    /// Note that we initialize with num_bins rather than bin_size
+    pub fn new(min: f64, max: f64, n_bins: usize) -> Result<Self, &'static str> {
+        if n_bins == 0 {
+            Err("Number of bins must be greater than zero")
+        } else if max <= min {
+            Err("Maximum value must be greater than minimum value")
+        } else if !min.is_finite() || !max.is_finite() {
+            Err("Min and max values must be finite")
+        } else {
+            Ok(Self {
+                min,
+                max,
+                bin_size: (max - min) / n_bins as f64,
+                n_bins,
+            })
+        }
+    }
+}
+
+impl BinEdges for RegularBinEdges {
+    fn bin_index(&self, value: f64) -> Option<usize> {
+        if value < self.min || value >= self.max {
+            return None;
+        }
+
+        // this cast handles the truncation
+        let index = ((value - self.min) / self.bin_size) as usize;
+
+        Some(index)
+    }
+
+    fn len(&self) -> usize {
+        self.n_bins
+    }
+}
+
+pub struct IrregularBinEdges<'a> {
+    bin_edges: &'a [f64],
+}
+
+impl<'a> IrregularBinEdges<'_> {
+    pub fn new(bin_edges: &'a [f64]) -> Result<IrregularBinEdges<'a>, &'static str> {
+        if bin_edges.len() < 2 {
+            return Err("A minimum of two bin edges are required");
+        }
+
+        // Check that all bin_edges are finite
+        // It may be worth supporting -inf and +inf as first and last bin edges
+
+        if bin_edges.iter().any(|&x| !x.is_finite()) {
+            return Err("Bin edges must be finite");
+        }
+
+        // Check if bin_edges are in strictly increasing order
+        for i in 1..bin_edges.len() {
+            if bin_edges[i] <= bin_edges[i - 1] {
+                return Err("Bin edges must be in strictly increasing order");
+            }
+        }
+
+        Ok(IrregularBinEdges { bin_edges })
+    }
+}
+
+impl BinEdges for IrregularBinEdges<'_> {
+    fn bin_index(&self, value: f64) -> Option<usize> {
+        if value < self.bin_edges[0] || value >= self.bin_edges[self.bin_edges.len() - 1] {
+            return None;
+        }
+
+        let index = self
+            .bin_edges
+            // There may be downsides to using total_cmp (perf?)
+            .binary_search_by(|probe| probe.total_cmp(&value))
```

Comment:
That's nifty! I didn't know this method existed!

Aside: when we get around to writing SIMD, I suspect a linear search will be faster, but let's cross that bridge when we get to it